From 2099262dcbca6a5ee703f41f2cdac88535a2ed11 Mon Sep 17 00:00:00 2001
From: Maharaja Kennadyrajan <quic_mkenna@quicinc.com>
Date: Wed, 13 Dec 2023 11:24:49 +0530
Subject: [PATCH] wifi: ath12k: Add debugfs support to configure primary link

Add a vif specific debugfs entry to select the primary link to be used by
the user in MLO case for the AP mode.

This will helpful while running the functionality and performance test
case.

echo N > /sys/kernel/debug/ieee80211/phy1/netdev:wlan0/primary_link

where N = 0 to set link-0 as primary link, 1 to set link-1 as primary
link and 2 to set link-2 as primary link.

cat /sys/kernel/debug/ieee80211/phy1/netdev:wlan0/primary_link to
get the configured primary link.

Co-developed-by: Avula Sri Charan <quic_asrichar@quicinc.com>
Signed-off-by: Avula Sri Charan <quic_asrichar@quicinc.com>
Signed-off-by: Maharaja Kennadyrajan <quic_mkenna@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h    |  1 +
 drivers/net/wireless/ath/ath12k/debugfs.c | 74 +++++++++++++++++++----
 drivers/net/wireless/ath/ath12k/debugfs.h |  5 +-
 drivers/net/wireless/ath/ath12k/mac.c     |  4 +-
 4 files changed, 66 insertions(+), 18 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -487,8 +487,10 @@ struct ath12k_vif {
 	spinlock_t mcbc_gsn_lock; /* Protect mcbc_gsn */
 	struct ath12k_vif_cache cache[IEEE80211_MLD_MAX_NUM_LINKS];
 	u8 last_scan_link;
+	u8 primary_link_id;
 #ifdef CPTCFG_ATH12K_DEBUGFS
 	struct dentry *debugfs_linkstats;
+	struct dentry *debugfs_primary_link;
 #endif /* CPTCFG_ATH12K_DEBUGFS */
 	bool mcast_to_ucast_en;
  };
--- a/drivers/net/wireless/ath/ath12k/debugfs.c
+++ b/drivers/net/wireless/ath/ath12k/debugfs.c
@@ -4818,22 +4818,70 @@ static const struct file_operations ath1
 	.llseek = default_llseek,
 };
 
-int ath12k_debugfs_add_interface(struct ath12k_link_vif *arvif)
+static ssize_t ath12k_write_primary_link(struct file *file,
+					 const char __user *user_buf,
+					 size_t count, loff_t *ppos)
+{
+	struct ath12k_vif *ahvif = file->private_data;
+	struct ath12k_hw *ah = ahvif->ah;
+	struct ath12k *ar = ah->radio;
+	u8 primary_link;
+
+	if (kstrtou8_from_user(user_buf, count, 0, &primary_link))
+		return -EINVAL;
+
+	if (!(ahvif->links_map & BIT(primary_link))) {
+		ath12k_warn(ar->ab, "Invalid link id : %u\n", primary_link);
+		return -EINVAL;
+	}
+
+	mutex_lock(&ah->conf_mutex);
+	ahvif->primary_link_id = primary_link;
+	mutex_unlock(&ah->conf_mutex);
+
+	return count;
+}
+
+static ssize_t ath12k_read_primary_link(struct file *file,
+					char __user *ubuf,
+					size_t count, loff_t *ppos)
+{
+	struct ath12k_vif *ahvif = file->private_data;
+	struct ath12k_hw *ah = ahvif->ah;
+	int len = 0;
+	char buf[32] = {0};
+
+	mutex_lock(&ah->conf_mutex);
+	len = scnprintf(buf, sizeof(buf) - len, "Primary link_id: %u\n",
+			ahvif->primary_link_id);
+	mutex_unlock(&ah->conf_mutex);
+
+	return simple_read_from_buffer(ubuf, count, ppos, buf, len);
+}
+
+static const struct file_operations ath12k_fops_primary_link = {
+	.open = simple_open,
+	.write = ath12k_write_primary_link,
+	.read = ath12k_read_primary_link,
+	.owner = THIS_MODULE,
+	.llseek = default_llseek,
+};
+
+void ath12k_debugfs_add_interface(struct ath12k_link_vif *arvif)
 {
 	struct ath12k_vif *ahvif = arvif->ahvif;
 	struct ieee80211_hw *hw = arvif->ar->ah->hw;
 	struct ieee80211_vif *vif = ahvif->vif;
 	u8 link_id = arvif->link_id;
-	int ret = 0;
 
 	if (link_id >= IEEE80211_MLD_MAX_NUM_LINKS)
-		goto exit;
+		return;
 
 	if (ahvif->vif->type != NL80211_IFTYPE_AP)
-		goto exit;
+		return;
 
 	if (arvif->debugfs_twt)
-		goto exit;
+		return;
 
 	if (!hw->wiphy->num_hw) {
 		arvif->debugfs_twt = debugfs_create_dir("twt",
@@ -4853,8 +4901,7 @@ int ath12k_debugfs_add_interface(struct
 			    "failed to create directory %p\n",
 			    arvif->debugfs_twt);
 		arvif->debugfs_twt = NULL;
-		ret = -1;
-		goto exit;
+		return;
 	}
 
 	debugfs_create_file("add_dialog", 0200, arvif->debugfs_twt,
@@ -4877,13 +4924,12 @@ int ath12k_debugfs_add_interface(struct
 			debugfs_remove_recursive(arvif->debugfs_twt);
 			arvif->debugfs_twt = NULL;
 			arvif->debugfs_rtsthreshold = NULL;
-			ret = -ENOMEM;
-			goto exit;
+			return;
 		}
 	}
 
 	if (ahvif->debugfs_linkstats)
-		goto exit;
+		return;
 
 	ahvif->debugfs_linkstats = debugfs_create_file("link_stats",
 						       0200,
@@ -4905,8 +4951,11 @@ int ath12k_debugfs_add_interface(struct
 		ahvif->debugfs_linkstats = NULL;
 	}
 
-exit:
-	return ret;
+	ahvif->debugfs_primary_link = debugfs_create_file("primary_link",
+							  0644,
+							  vif->debugfs_dir,
+							  ahvif,
+							  &ath12k_fops_primary_link);
 }
 
 void ath12k_debugfs_remove_interface(struct ath12k_link_vif *arvif)
@@ -4919,9 +4968,13 @@ void ath12k_debugfs_remove_interface(str
 		arvif->debugfs_twt = NULL;
 		arvif->debugfs_rtsthreshold = NULL;
 		ahvif->debugfs_linkstats = NULL;
+		ahvif->debugfs_primary_link = NULL;
 		return;
 	}
 
+	debugfs_remove(ahvif->debugfs_primary_link);
+	ahvif->debugfs_primary_link = NULL;
+
 	if (ahvif->debugfs_linkstats) {
 		debugfs_remove(ahvif->debugfs_linkstats);
 		ahvif->debugfs_linkstats = NULL;
--- a/drivers/net/wireless/ath/ath12k/debugfs.h
+++ b/drivers/net/wireless/ath/ath12k/debugfs.h
@@ -310,7 +310,7 @@ static inline int ath12k_debugfs_rx_filt
 	return ar->debug.rx_filter;
 }
 
-int ath12k_debugfs_add_interface(struct ath12k_link_vif *arvif);
+void ath12k_debugfs_add_interface(struct ath12k_link_vif *arvif);
 void ath12k_debugfs_remove_interface(struct ath12k_link_vif *arvif);
 
 #else
@@ -389,9 +389,8 @@ static inline int ath12k_debugfs_rx_filt
 	return 0;
 }
 
-static inline int ath12k_debugfs_add_interface(struct ath12k_link_vif *arvif)
+static inline void ath12k_debugfs_add_interface(struct ath12k_link_vif *arvif)
 {
-	return 0;
 }
 
 static inline void ath12k_debugfs_remove_interface(struct ath12k_link_vif *arvif)
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -10906,9 +10906,7 @@ static void ath12k_update_obss_color_not
 	}
 #endif
 
-	ret = ath12k_debugfs_add_interface(arvif);
-	if (ret)
-		goto err_peer_del;
+	ath12k_debugfs_add_interface(arvif);
 
 	return ret;
 
